sensors/sensors.c

```diff
@@ -0,0 +1,600 @@
+
+
+/* Update data from sensors */
+
+int sensors_publish(unsigned int devId, const sensor_event_t *event)
+{
+	uint8_t id = __builtin_ffs(event->type);
+	uint8_t evtId = sensors_common.devEvents[devId][id];
+
+	/*__builtin_ffs returns one plus the index of the least significant, otherwise 0 */
+	if (id == 0) {
+		return -EINVAL;
+	}
+
+	--id;
+	if (id >= NB_SENSOR_TYPES || sensors_common.events[id] == NULL || sensors_common.evtNb[id] < evtId) {
+		return -EINVAL;
+	}
+
+	mutexLock(sensors_common.locks[id]);
+	memcpy(&sensors_common.events[id][evtId], event, sizeof(sensor_event_t));
+	mutexUnlock(sensors_common.locks[id]);
+
+	return EOK;
+}
+
+
+/* Handle messages */
+
+static int sensors_open(void)
+{
+	int res;
+	sensor_client_t *client = calloc(1, sizeof(sensor_client_t));
+
+	if (client == NULL) {
+		return -ENOMEM;
+	}
+
+	mutexLock(sensors_common.cLock);
+	res = idtree_alloc(&sensors_common.clients, &client->node);
+	if (res < 0) {
+		free(client);
+	}
+	else {
+		res = CLIENT_SET_ID(res);
+	}
+	mutexUnlock(sensors_common.cLock);
+
+	return res;
+}
+
+
+static int sensors_close(id_t id)
+{
+	int err = -EINVAL;
+	sensor_client_t *client;
+
+	mutexLock(sensors_common.cLock);
+	client = lib_treeof(sensor_client_t, node, idtree_find(&sensors_common.clients, id));
+	if (client != NULL) {
+		err = EOK;
+		idtree_remove(&sensors_common.clients, &client->node);
+		free(client);
+	}
+	mutexUnlock(sensors_common.cLock);
+
+	return err;
+}
+
+
+static ssize_t sensors_read(sensors_data_t *data, size_t sz, id_t clientID)
+{
+	uint8_t id;
+	ssize_t res = -EINVAL;
+	sensor_type_t types;
+	size_t tempSz = 0, evtNb, chunkSz;
+	sensor_client_t *client;
+	sensor_event_t *events;
+
+	if (data == NULL) {
+		return res;
+	}
+
+	if (sz == 0) {
+		return 0;
+	}
+
+	tempSz = sizeof(data->size);
+	events = data->events;
+
+	mutexLock(sensors_common.cLock);
+	client = lib_treeof(sensor_client_t, node, idtree_find(&sensors_common.clients, clientID));
+	mutexUnlock(sensors_common.cLock);
+
+	if (client != NULL) {
+		types = client->ops.types;
+		/* Iterate only through available sensors for a client */
+		for (id = __builtin_ffs(types); id != 0; types &= ~(1 << (id - 1)), id = __builtin_ffs(types)) {
+			evtNb = sensors_common.evtNb[id - 1];
+			chunkSz = evtNb * sizeof(sensor_event_t);
+			if (tempSz + chunkSz > sz) {
+				break;
+			}
+
+			mutexLock(sensors_common.locks[id - 1]);
+			memcpy(events, sensors_common.events[id - 1], chunkSz);
+			mutexUnlock(sensors_common.locks[id - 1]);
+
+			events += evtNb;
+			tempSz += chunkSz;
+		}
+
+		data->size = (tempSz - sizeof(data->size)) / sizeof(sensor_event_t);
+		res = tempSz;
+	}
+
+	return res;
+}
+
+
+/* Function return total number of events for selected types */
+static unsigned int sensors_evtsCnt(sensor_type_t types)
+{
+	uint8_t id;
+	unsigned int sz = 0;
+
+	for (id = __builtin_ffs(types); id != 0; types &= ~(1 << (id - 1)), id = __builtin_ffs(types)) {
+		sz += sensors_common.evtNb[id - 1];
```

Comment:
Shouldn't this be protected by mutex?

Reply:
This table is only modified within init functions, after that it is read only.